src/librustc_trans/trans/adt.rs

```diff
@@ -462,36 +503,38 @@ fn find_discr_field_candidate<'tcx>(tcx: &ty::ctxt<'tcx>,
             match field_ty.sty {
                 ty::TyRawPtr(ty::TypeAndMut { ty, .. }) if !type_is_sized(tcx, ty) => {
                     path.extend_from_slice(&[0, FAT_PTR_ADDR]);
-                    Some(path)
+                    Some((path, C_null(type_of::sizing_type_of(cx, ty))))
                 },
                 ty::TyRawPtr(..) | ty::TyInt(..) | ty::TyUint(..) => {
                     path.push(0);
-                    Some(path)
+                    Some((path, C_null(type_of::sizing_type_of(cx, ty))))
```

Comment:
So, the forbidden value will be compared against the discriminant field. Say we had `Option<NonZero<int>>`, we're testing that that nested int is non-zero. Currently, it's returning `C_null` of `NonZero` (the passed in `ty` to the function) as the forbidden value rather than of the int field (`field_ty`).